File: sdk/java/src/main/java/dev/udb/client/UdbAssetClient.java
package dev.udb.client;

import com.udb.core.asset.services.v1.AssetServiceGrpc;
import com.udb.core.asset.services.v1.CompleteStepRequest;
import com.udb.core.asset.services.v1.CompleteStepResponse;
import com.udb.core.asset.services.v1.CreatePipelineDefinitionRequest;
import com.udb.core.asset.services.v1.CreatePipelineDefinitionResponse;
import com.udb.core.asset.services.v1.GetAssetRequest;
import com.udb.core.asset.services.v1.GetAssetResponse;
import com.udb.core.asset.services.v1.GetPipelineDefinitionRequest;
import com.udb.core.asset.services.v1.GetPipelineDefinitionResponse;
import com.udb.core.asset.services.v1.GetPipelineRequest;
import com.udb.core.asset.services.v1.GetPipelineResponse;
import com.udb.core.asset.services.v1.ListAssetsRequest;
import com.udb.core.asset.services.v1.ListAssetsResponse;
import com.udb.core.asset.services.v1.RegisterAssetRequest;
import com.udb.core.asset.services.v1.RegisterAssetResponse;
import com.udb.core.asset.services.v1.StartPipelineRequest;
import com.udb.core.asset.services.v1.StartPipelineResponse;
import io.grpc.Channel;

/**
 * Blocking facade over the native {@code AssetService} (asset registration +
 * processing pipelines). Rides the shared control-plane channel and attaches the
 * project {@link UdbMetadata} headers to every call. The raw generated stub stays
 * reachable via {@link #stub()}.
 */
public final class UdbAssetClient {
  private final AssetServiceGrpc.AssetServiceBlockingStub stub;

  UdbAssetClient(Channel channel, UdbMetadata metadata) {
    this(channel, metadata, UdbCredentials.fromMetadata(metadata));
  }

  UdbAssetClient(Channel channel, UdbMetadata metadata, UdbCredentials credentials) {
    this.stub =
        AssetServiceGrpc.newBlockingStub(channel)
            .withInterceptors(UdbClient.credentialInterceptor(metadata, credentials));
  }

  /** The raw generated blocking stub (never hidden). */
  public AssetServiceGrpc.AssetServiceBlockingStub stub() {
    return stub;
  }

  /** Define a reusable processing pipeline (ordered steps). */
  public CreatePipelineDefinitionResponse createPipelineDefinition(
      CreatePipelineDefinitionRequest request) {
    return stub.createPipelineDefinition(request);
  }

  /** Fetch a pipeline definition by id. */
  public GetPipelineDefinitionResponse getPipelineDefinition(
      GetPipelineDefinitionRequest request) {
    return stub.getPipelineDefinition(request);
  }

  /** Register an asset (typically backed by a stored file). */
  public RegisterAssetResponse registerAsset(RegisterAssetRequest request) {
    return stub.registerAsset(request);
  }

  /** Start a pipeline instance over an asset. */
  public StartPipelineResponse startPipeline(StartPipelineRequest request) {
    return stub.startPipeline(request);
  }

  /** Fetch a running/finished pipeline instance by id. */
  public GetPipelineResponse getPipeline(GetPipelineRequest request) {
    return stub.getPipeline(request);
  }

  /** Mark a pipeline step complete (advances the instance). */
  public CompleteStepResponse completeStep(CompleteStepRequest request) {
    return stub.completeStep(request);
  }

  /** List assets for the tenant, with paging/filter on the request. */
  public ListAssetsResponse listAssets(ListAssetsRequest request) {
    return stub.listAssets(request);
  }

  /** Fetch an asset's metadata by id. */
  public GetAssetResponse getAsset(GetAssetRequest request) {
    return stub.getAsset(request);
  }
}

File: sdk/csharp/Udb.Client/UdbAssetClient.cs
using Grpc.Core;
using AssetV1 = Udb.Core.Asset.Services.V1;

namespace Udb.Client;

/// <summary>
/// Ergonomic async facade over the native <c>AssetService</c> (asset registration
/// and processing pipelines). Each wrapper applies the shared <see cref="UdbProject"/>
/// metadata headers and returns the unwrapped response task. The raw generated
/// client stays reachable via <see cref="Raw"/>.
/// </summary>
public sealed class UdbAssetClient
{
    private readonly AssetV1.AssetService.AssetServiceClient _client;
    private readonly Func<Metadata> _headers;

    internal UdbAssetClient(AssetV1.AssetService.AssetServiceClient client, Func<Metadata> headers)
    {
        _client = client;
        _headers = headers;
    }

    /// <summary>Raw generated asset service client.</summary>
    public AssetV1.AssetService.AssetServiceClient Raw => _client;

    /// <summary>Create a reusable pipeline definition.</summary>
    public Task<AssetV1.CreatePipelineDefinitionResponse> CreatePipelineDefinitionAsync(
        AssetV1.CreatePipelineDefinitionRequest request, CancellationToken ct = default)
        => _client.CreatePipelineDefinitionAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Get a pipeline definition.</summary>
    public Task<AssetV1.GetPipelineDefinitionResponse> GetPipelineDefinitionAsync(
        AssetV1.GetPipelineDefinitionRequest request, CancellationToken ct = default)
        => _client.GetPipelineDefinitionAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Register a managed asset wrapping a storage file.</summary>
    public Task<AssetV1.RegisterAssetResponse> RegisterAssetAsync(
        AssetV1.RegisterAssetRequest request, CancellationToken ct = default)
        => _client.RegisterAssetAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Start a pipeline instance for an asset.</summary>
    public Task<AssetV1.StartPipelineResponse> StartPipelineAsync(
        AssetV1.StartPipelineRequest request, CancellationToken ct = default)
        => _client.StartPipelineAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Get a pipeline instance with its steps.</summary>
    public Task<AssetV1.GetPipelineResponse> GetPipelineAsync(
        AssetV1.GetPipelineRequest request, CancellationToken ct = default)
        => _client.GetPipelineAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Complete (or skip/fail) a pipeline step.</summary>
    public Task<AssetV1.CompleteStepResponse> CompleteStepAsync(
        AssetV1.CompleteStepRequest request, CancellationToken ct = default)
        => _client.CompleteStepAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>List assets.</summary>
    public Task<AssetV1.ListAssetsResponse> ListAssetsAsync(
        AssetV1.ListAssetsRequest request, CancellationToken ct = default)
        => _client.ListAssetsAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Get an asset.</summary>
    public Task<AssetV1.GetAssetResponse> GetAssetAsync(
        AssetV1.GetAssetRequest request, CancellationToken ct = default)
        => _client.GetAssetAsync(request, _headers(), cancellationToken: ct).ResponseAsync;
}
